django/template/base.py

```diff
@@ -393,28 +393,34 @@ def create_token(self, token_string, position, lineno, in_tag):
 
 
 class DebugLexer(Lexer):
+    # This parallels the use of tag_re.split() in Lexer.tokenize().
+    def _tag_re_split(self):
+        prev_end = 0
+        for match in tag_re.finditer(self.template_string):
+            start, end = match.span()
+            yield (prev_end, start)
+            yield (start, end)
+            prev_end = end
+        yield (prev_end, len(self.template_string))
+
     def tokenize(self):
         """
         Split a template string into tokens and annotates each token with its
         start and end position in the source. This is slower than the default
         lexer so only use it when debug is True.
         """
+        # For maintainability, it is helpful if the implementation below can
+        # continue to closely parallel Lexer.tokenize()'s implementation.
+        template_string = self.template_string
+        in_tag = False
         lineno = 1
         result = []
-        upto = 0
-        for match in tag_re.finditer(self.template_string):
-            start, end = match.span()
-            if start > upto:
-                token_string = self.template_string[upto:start]
-                result.append(self.create_token(token_string, (upto, start), lineno, in_tag=False))
+        for position in self._tag_re_split():
+            token_string = template_string[slice(*position)]
```

Comment:
I think that you could move the slicing inside `DebugLexer._tag_re_split()` and then `DebugLexer.tokenize()` will be even closer to `Lexer.tokenize()`:

```suggestion
        for bit, position in self._tag_re_split():
```

Maybe with these changes it makes sense to rename `DebugLexer._tag_re_split()` to something like `.split()` and add the same method to `Lexer` with something like:

```python
    def split():
        yield from ((bit, None) for bit in tag_re.split(self.template_string))
```

Then you should be able to ditch `DebugLexer.tokenize()` entirely and inherit it.

Reply:
Yeah, that had occurred to me, too. But I think it would be better if we didn't let the debug implementation slow the non-debug implementation more.

Reply:
I did move the slicing to inside `DebugLexer._tag_re_split()` though like you said.

Reply:
> But I think it would be better if we didn't let the debug implementation slow the non-debug implementation more.

Yeah. That did occur to me, but haven't had time to check the performance. It might still work out in concert with other ideas I've had… But those don't need to hold this up.